### ISSUE TITLE
Idea: It would be handy to have an Events component

### ISSUE THREAD

**Issue by chrispauly** [exceptionless/Foundatio#12]
Similar to the logging and metrics (probably), it would be useful to have an eventqueue that all of your classes/methods publish to (in the situations where someone initializes it).  Then subscribers of the event queue could listen for certain events and perform actions based on it.

This concept could basically leverage existing components you have already.  And I haven't looked at your all of your code yet, but maybe this is how your metrics already works.

The use case for this would be something like a Job that subscribes for all SaveFile events... and as soon as it happens, it runs a command.  But that same Job may run for DeleteFile events.  And so, it can just subscribe to that event too.  I know this feature is something that easily could be built into a custom project utilizing other foundatio components.  However, I think it would be great to be a foundational pipeline used by your framework.

**Comment by niemyjski** [comment]
@ejsmith: Could we do this with the new attached behaviors?

**Comment by ejsmith** [comment]
I think you would want to use the messagebus for this.  You could publish a message from wherever you are saving the file and then you could subscribe to that event and enqueue an action to happen or just do the action right then.  Does that make sense?

**Comment by chrispauly** [comment]
I was thinking leveraging the messagebus would make the most sense too... since it already has the pub/sub functionality and ability to perform an action based on an object type (which would be the specific EventArgs message for whatever event occurred).  This architecture seems like a way you could provide metrics across every function (as long as they implement the event messaging).

Also, it's a way that I've used in the past to handle asynchronous processing of synchronous tasks cross-processes.

**Comment by niemyjski** [comment]
@chrispauly did you end up going with the message bus approach? After rereading all of this it seems like that is what you were wanting to use. Is this correct? If not please let us know. Also, please feel free to close this issue if you feel this use case is solved.

**Comment by chrispauly** [comment]
Oh, I never did anything with this.  It was just an idea I had after you presented at the MadDotNet meeting last year.  I really liked this collection and just thought that Eventing was something that I leverage in many apps I built and thought it would fit right into a foundational collection.  I've changed jobs since last year and haven't written any new custom apps that could use these objects, but if I do find the need I will definitely look at forking and/or creating a pull request.
